payjoin/src/core/uri/url_ext.rs

```diff
@@ -110,22 +107,57 @@ pub fn parse_with_fragment(endpoint: &str) -> Result<Url, BadEndpointError> {
     Ok(url)
 }
 
-fn get_param<F, T>(url: &Url, prefix: &str, parse: F) -> Option<T>
-where
-    F: Fn(&str) -> Option<T>,
-{
-    if let Some(fragment) = url.fragment() {
-        let mut delim = '-';
-
-        // For backwards compatibility, also accept `+` as a
-        // fragment parameter delimiter. This was previously
-        // specified, but may be interpreted as ` ` by some
-        // URI parsoing libraries. Therefore if `-` is missing,
-        // assume the URI was generated following the older
-        // version of the spec.
-        if !fragment.contains(delim) {
-            delim = '+';
+#[derive(Debug)]
+enum ParseFragmentError {
+    InvalidChar(char),
+    AmbiguousDelimiter,
+}
+
+fn check_fragment_delimiter(fragment: &str) -> Result<char, ParseFragmentError> {
+    // For backwards compatibility, also accept `+` as a
+    // fragment parameter delimiter. This was previously
+    // specified, but may be interpreted as ` ` by some
+    // URI parsoing libraries. Therefore if `-` is missing,
+    // assume the URI was generated following the older
+    // version of the spec.
+    if fragment.is_empty() {
+        return Ok('-');
+    }
+
+    let mut has_plus = false;
+    let mut has_dash = false;
+
+    // Even though fragment is a &str, it should be ascii so bytes() correspond
+    // to chars(), except that it's easier to check that they are in range
+    for c in fragment.bytes() {
+        // check for allowed delimiters
+        if c == b'-' {
+            has_dash = true;
+        } else if c == b'+' {
+            has_plus = true;
+        }
```

Comment:
Does this need to be in the loop? Can't fragment.contains('-') and fragment.contains('+') be used outside of the loop in the match for greater legibility

Reply:
yeah I understand why you'd put it all together now and am comfortable with that though would merge this as-is at this point.